### PR TITLE
VS compile fixes

### PR DESCRIPTION
**Pull request by lab313ru**
Should fix all compilation errors in Visual Studio.

### CODE REVIEW

**Comment by nemerle** [comment]
'and' 'not' 'or' are standard keywords that are not supported by MSVC in default mode.
instead of replacing them, you could use the compatibility header Microsoft added to 'fix' the problem.

``` cpp
#ifdef _MSC_VER
#include <iso646.h>
#endif
```

**Comment by lab313ru** [comment]
Anyway mixing 'not' and '!', 'and' and '&&' is not a good practice. So, instead of adding such headers, which may not be available in some vs versions, let it be more correct.

**Comment by nemerle** [comment]
Ok, the issue with 'and', 'or' 'not' keywords should be patched up now. And those keywords should be used consistently throughout the codebase.

**Comment by lab313ru** [comment]
There are things, that still does not compatible with Visual Studio.
Could you merge some more fixes from my request?

File: src/scanner.cpp [hunk 1]
```diff
@@ -5,6 +5,7 @@
  * (C) Cristina Cifuentes, Jeff Ledermann
  ****************************************************************************/
 
+#define NOMINMAX
```
Comment (nemerle) [inline src/scanner.cpp:8]: Is this part of root CMakeLists.txt not solving this ?

``` cmake
IF(CMAKE_BUILD_TOOL MATCHES "(msdev|devenv|nmake)")
    ADD_DEFINITIONS(-D_CRT_SECURE_NO_WARNINGS -D__UNIX__ -D_CRT_NONSTDC_NO_DEPRECATE -DNOMINMAX)
    ADD_DEFINITIONS(/W4)
ELSE()
```

Comment (lab313ru) [inline src/scanner.cpp:8]: It should fix that problem.

**Comment by nemerle** [comment]
I think we could consider removing LLVM as a dependency for now, it makes compiling things troublesome on windows, and no interesting LLVM  features are used in the codebase anyway :)

**Comment by lab313ru** [comment]
Good idea. And then, we'll try to fix rest of errors.

**Comment by nemerle** [comment]
The latest commit (58532f4) should make dcc llvm-free

File: src/scanner.cpp (hunk 1 above)
Comment (lab313ru) [inline src/scanner.cpp:8]: scanner.cpp and ast.cpp still have problem with std::min/max. VS tries to use non-std version.

They are in dcc_lib project. There is no such define in this proect.

**Comment by lab313ru** [comment]
Allmost good! But, problem with **std::min/max** still exists.
And **Boost** (by **dcc_original** project) requires linking with (there is no path to this library in cmake).
**libboost_regex-vc140-mt-sgd-1_60.lib**

**Comment by nemerle** [comment]
The last commit should have fixed the std::min/max problem.
As for boost, we don't use boost_regex in the project ?

**Comment by lab313ru** [comment]
I thought no. But VS requires it. I don't know why. I even not found any requirement for this file anywhere.

**Comment by nemerle** [comment]
hmm...  we use header-only boost libraries, I have no idea why your compilation would try to use regex ?
Maybe some kind of boost configuration issue ?

**Comment by lab313ru** [comment]
Maybe. I think I have to rebuild it.

**Comment by nemerle** [comment]
Nope, I've managed to replicate this... investigating

**Comment by nemerle** [comment]
Found the problem, fix soon

**Comment by nemerle** [comment]
Ok, fixes are commited. The dcc builds on msvc2015

**Comment by lab313ru** [comment]
And, last things:
- is to setup **/W3** option by default, but not **/Wall** for warnings.
- Set **/MTd** and **/MT** options to use static CRT linkinage.

**Comment by nemerle** [comment]
Well setting /W4 in the CMakeLists.txt is done to encourage fixing those warnings ;)

And static linkage.. that's another issue - why would we do that ?

Will close this PR, since the msvc2015 build 'works' now ?

**Comment by lab313ru** [comment]
Yep, it works now.

**Comment by lab313ru** [comment]
Ups, sorry.

```
for(uint32_t v : prog.relocTable) { // <- iterator is not dereferensable in DccFrontend.cpp
```

**Comment by lab313ru** [comment]
I had not only compiling fixes, but launching fixes too.

**Comment by nemerle** [comment]
That range-based for loop _should_ be working.
I have no idea yet why msvc barfs on this code.

**Comment by lab313ru** [comment]
But it happens.
Could you rewrite it as classic iterator loop?

**Comment by nemerle** [comment]
Found the actual bug, fix is in the latest commit

**Comment by lab313ru** [comment]
Ok. Looks like old bugs are fixed now.

**Pull request closed without merging**